0017-ilp-errors/0017-ilp-errors.md

```diff
@@ -0,0 +1,73 @@
+# Interledger Protocol Errors
+
+This spec defines a standard error format and codes to be used when ILP payments are rejected.
+
+Inspired by [HTTP Status Codes](https://tools.ietf.org/html/rfc2616#section-10), ILP errors are categorized based on the intended behavior of the caller when they get the given error.
+
+## Error Propagation
+
+ILP connectors MUST relay errors when they are notified of a transfer being rejected.
+
+Connectors SHOULD NOT modify the error codes or messages except in the following way: connectors MAY append their ILP addresses, separated from the error message by a newline (`\n`), to assist with debugging/tracing.
+
+## Data Format
+
+ILP errors are UTF-8 encoded strings that are a maximum of 65535 bytes long. Error messages that are too long for a ledger to handle SHOULD be truncated to the first 65535 bytes.
+
+Errors consist of a three character code, followed by a colon (`:`) and an arbitrary message set by the entity that raised the error. Transport and application layer protocols MAY define error messages that have special significance in their context.
+
+```
+<Error Code>: <Error Message>
+<Relaying Connector ILP Address>
+<Relaying Connector ILP Address>
+...
+```
+
+## Error Codes
+
+### S__ - Sender Error
+
+Sender errors indicate that the payment is invalid and should not be retried unless the details are changed.
+
+| Code | Name | Description |
+|---|---|---|
+| **S00** | **Bad Request** | Generic sender error. |
+| **S01** | **Invalid Packet** | The ILP packet was syntactically invalid. |
+| **S02** | **Unreachable** | There was no way to forward the payment, because the destination ILP address was wrong or unreachable. |
+| **S03** | **Invalid Amount** | The amount is invalid, e.g. it contains more digits of precision than are available on the destination ledger or the amount is greater than the total amount of the given asset in existence. |
+| **S04** | **Insufficient Destination Amount** | The amount was insufficient (e.g. you tried to pay a $100 invoice with $10). |
```

Comment:
S04 is only for the receiver to emit and means "the packet and transfer got here just fine, but it's less than you need to pay me for me to accept it".
S03 means the amount was just invalid
R02 is used when the individual transfer amount isn't enough for the connector to pass it on